src/uct/tcp/tcp_sockcm.c

```diff
@@ -54,33 +54,56 @@ static void uct_tcp_close_ep(uct_tcp_sockcm_ep_t *ep)
 void uct_tcp_sa_data_handler(int fd, void *arg)
 {
     uct_tcp_sockcm_ep_t *ep = (uct_tcp_sockcm_ep_t*)arg;
+    ucs_status_t status;
 
     ucs_assertv(ep->fd == fd, "ep->fd %d fd %d, ep_state %d", ep->fd, fd, ep->state);
 
-    if (ep->state & UCT_TCP_SOCKCM_EP_ON_SERVER) {
-        /* server side - the client disconnected and the server is waiting
-         * for EVREAD events */
-        uct_tcp_close_ep(ep);
-        return;
-    }
-
     if (!ucs_socket_is_connected(fd)) {
-        ucs_debug("fd %d is not connected", fd);
+        ucs_debug("fd %d is not connected. ep state: %d", fd, ep->state);
         /* coverity[check_return] */
         ucs_async_modify_handler(fd, 0);
         return;
     }
 
     switch (ep->state) {
     case UCT_TCP_SOCKCM_EP_ON_CLIENT:
-        /* connect() completed */
+        /* connect() completed, send data to the server */
         ep->state |= UCT_TCP_SOCKCM_EP_CONNECTED;
-        /* TODO: start sending the user's data */
 
-        ucs_async_remove_handler(ep->fd, 1);
-        close(ep->fd);
-        ep->fd = -1;
+        status = uct_tcp_sockcm_ep_send_priv_data(ep, 1);
+        if (status != UCS_OK) {
+            ucs_async_modify_handler(fd, 0);
+        }
+        break;
+    case UCT_TCP_SOCKCM_EP_ON_CLIENT | UCT_TCP_SOCKCM_EP_CONN_SENDING:
+         /* can send, progress the sending */
+         status = uct_tcp_sockcm_ep_progress_send(ep, 1);
+         if (status != UCS_OK) {
+             ucs_async_modify_handler(fd, 0);
+         }
+         break;
+    case UCT_TCP_SOCKCM_EP_ON_CLIENT | UCT_TCP_SOCKCM_EP_CONN_SENT:
+         /* finished sending. TODO recv data from the server */
+         ucs_async_modify_handler(fd, 0);
+         break;
+    case UCT_TCP_SOCKCM_EP_ON_SERVER | UCT_TCP_SOCKCM_EP_CONNECTED:
+        /* receive data from the client */
+        status = uct_tcp_sockcm_ep_recv(ep);
+        if (status != UCS_OK) {
+            uct_tcp_close_ep(ep);
+        }
         break;
+    case UCT_TCP_SOCKCM_EP_ON_SERVER | UCT_TCP_SOCKCM_EP_CONN_RECEIVING:
+         /* can read, progress the receving */
+         status = uct_tcp_sockcm_ep_progress_recv(ep);
```

Comment:
in next PR, when server also has to send and client also has to receive, will we have 2x number of cases in this switch/case?

Reply:
i think we will have 4 more cases - server_sending, server_sent, client_receiving, client_received.
and a couple cases for disconnecting. so around the amount of rdmacm events we have in the switch-case in rdmacm_cm.c.

Reply:
so better check each flag separately. e.g if RECEIVING call progress_recv. if SENDING call progress_send. both on client and server

Reply:
can this wait for the next PR? because in this one some of the functions are still client or server oriented. i think that once the flow is completed, i can unite.